packages/patterns/src/store/actions.js

```diff
@@ -40,18 +34,43 @@ export const __experimentalCreatePattern =
 			.dispatch( coreStore )
 			.saveEntityRecord( 'postType', 'wp_block', reusableBlock );
 
-		if ( syncType === 'unsynced' || ! clientIds ) {
-			return updatedRecord;
+		return updatedRecord;
+	};
+
+/**
+ * Create a pattern from a JSON file.
+ * @param {File} file The JSON file instance of the pattern.
+ */
+export const __experimentalCreatePatternFromFile =
+	( file ) =>
+	async ( { dispatch } ) => {
+		const fileContent = await file.text();
+		/** @type {import('./types').PatternJSON} */
+		let parsedContent;
+		try {
+			parsedContent = JSON.parse( fileContent );
+		} catch ( e ) {
+			throw new Error( 'Invalid JSON file' );
+		}
+		if (
+			parsedContent.__file !== 'wp_block' ||
+			! parsedContent.title ||
+			! parsedContent.content ||
+			typeof parsedContent.title !== 'string' ||
+			typeof parsedContent.content !== 'string' ||
+			( parsedContent.syncStatus &&
+				typeof parsedContent.syncStatus !== 'string' )
+		) {
+			throw new Error( 'Invalid Pattern JSON file' );
 		}
 
-		const newBlock = createBlock( 'core/block', {
-			ref: updatedRecord.id,
-		} );
-		registry
-			.dispatch( blockEditorStore )
-			.replaceBlocks( clientIds, newBlock );
-		dispatch.__experimentalSetEditingPattern( newBlock.clientId, true );
-		return updatedRecord;
+		const pattern = await dispatch.__experimentalCreatePattern(
+			parsedContent.title,
```

Comment:
Question: Should we increment the title if there's a same title already in the patterns list?

Reply:
Ultimately, having unique pattern names would be a good thing although to date there hasn't been anything enforcing it. There's already an issue open for this in https://github.com/WordPress/gutenberg/issues/52360. 

Unique names would help when we come to introducing slugs so they are consistent across all types of patterns.

Reply:
I'm thinking if we ever want to enforce the uniqueness then it should be done in the backend. I guess it's better to leave it for future PRs. WDYT?

Reply:
Agreed. We'll focus on that in a follow and can discuss options via the dedicated issue.